### PR TITLE
[GOONETICS] Chromosomes

### LINKED ISSUE #42632: genetic stability is broken af
I took mutadone after getting message that my dna is exploding. i lost my powers but still got hit with aftereffect (it happened to be spitting out organs)

### PR DESCRIPTION
**Pull request by Time-Green**
Finally adds chromosomes for mutations.
What they do: They pretty much act as mutation boosters

Stabilizer: Decreases genetic damage
Synchronizer: Decreases chance to backfire or hurt you
Power: Increases strength of some mutations
Energy: Decreases cooldown of action mutations

Every mutation can be given one chromosome, assuming that it would work. You can't give a power booster to chronic cough for example, since extremely strong coughs is a dumb concept.
:cl: Time-Green
add: Adds chromosomes to mutations. You can now alter your mutations in unexpected ways
/:cl:

You gain chromosomes from giving people succesful activators. Right now they can just absolutely spam the shit out of them by activating and deactivating the mutations. I'm debating wheter I should just make every mutation only work once for each person.

Right now with chromosomes it just copies the statistics, if it's preferable to keep the chromosome around as an object and draw info from that, that's cool and I'll get on that. 

----
(ed. SpaceManiac) Fixes #42632.

### CODE REVIEW

**Comment by GuyonBroadway** [comment]
So we can get around to removing the cr23 every crewmember seems to have.

**Comment by Time-Green** [comment]
> So we can get around to removing the cr23 every crewmember seems to have.

we can only give them more chromosomes

**Comment by bigfatbananacyclops** [comment]
YES

File: code/game/machinery/computer/dna_console.dm [hunk 1]
```diff
@@ -335,7 +360,14 @@
 					temp_html += "<td><span class='linkOff'>Combine</span></td></tr>"
 				else
 					temp_html += "<td><a href='?src=[REF(src)];task=combine;num=[i]'>Combine</a></td></tr>"
+			temp_html += "</table><br>"
+			temp_html += "<h3>Chromosome Storage:<br></h3>"
+			temp_html += "<table>"
+			for(var/obj/item/chromosome/CM in stored_chromosomes)
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 			for(var/i in 1 to stored_chromosomes.len)
```

File: code/game/machinery/computer/dna_console.dm [hunk 2]
```diff
@@ -335,7 +360,14 @@
...
+				var/i = stored_chromosomes.Find(CM)
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]: var/obj/item/chromosome/CM = stored_chromosomes[i]

File: code/game/machinery/computer/dna_console.dm [hunk 3]
```diff
@@ -651,7 +690,7 @@
 							var/datum/mutation/human/A = new HM.type()
 							A.copy_mutation(HM)
 							succes = TRUE
-							stored_mutations[A] = get_sequence(mutation) //We only store active mutations and all active mutations have the full sequence.
+							stored_mutations.Add(A)
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 							stored_mutations += A
```

File: code/game/machinery/computer/dna_console.dm [hunk 4]
```diff
@@ -735,7 +774,7 @@
 					var/datum/mutation/human/A = diskette.mutations[num]
 					var/datum/mutation/human/HM = new A.type()
 					HM.copy_mutation(A)
-					stored_mutations[HM] = get_sequence(HM.type)
+					stored_mutations.Add(HM)
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 					stored_mutations += HM
```

File: code/game/machinery/computer/dna_console.dm [hunk 5]
```diff
@@ -745,7 +784,7 @@
 					if(combine)
 						var/result_path = get_mixed_mutation(combine, path)
 						if(result_path)
-							stored_mutations[new result_path()] = get_sequence(result_path)
+							stored_mutations.Add(new result_path())
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 							stored_mutations += new result_path()
```

File: code/game/machinery/computer/dna_console.dm [hunk 6]
```diff
@@ -757,9 +796,29 @@
 						to_chat(usr,"<span class='notice'>Selected [A.name] for combining</span>")
 				else
 					to_chat(usr, "<span class='warning'>Not enough space to store potential mutation.</span>")
+		if("ejectchromosome")
+			if(LAZYLEN(stored_chromosomes) <= num)
+				var/obj/item/chromosome/CM = stored_chromosomes[num]
+				CM.forceMove(drop_location())
+				adjust_item_drop_location(CM)
+				stored_chromosomes.Remove(CM)
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 				stored_chromosomes -= CM
```

File: code/game/machinery/computer/dna_console.dm [hunk 7]
```diff
@@ -757,9 +796,29 @@
...
+				var/datum/mutation/human/HM = viable_occupant.dna.mutations[num]
+				var/list/chromosomes = list()
+				for(var/obj/item/chromosome/CM in stored_chromosomes)
+					if(CM.can_apply(HM))
+						chromosomes.Add(CM)
+				if(LAZYLEN(chromosomes))
+					var/obj/item/chromosome/CM = input("Select a chromosome to apply", "Apply Chromosome") as null|anything in chromosomes
+					if(CM)
+						to_chat(usr, "<span class='notice'>You apply [CM] to [HM.name].")
+						stored_chromosomes.Remove(CM)
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 						stored_chromosomes -= CM
```

File: code/game/machinery/computer/dna_console.dm [hunk 8]
```diff
@@ -757,9 +796,29 @@
...
+				CM.forceMove(drop_location())
+				adjust_item_drop_location(CM)
+				stored_chromosomes.Remove(CM)
+		if("applychromosome")
+			if(viable_occupant && (LAZYLEN(viable_occupant.dna.mutations) <= num))
+				var/datum/mutation/human/HM = viable_occupant.dna.mutations[num]
+				var/list/chromosomes = list()
+				for(var/obj/item/chromosome/CM in stored_chromosomes)
+					if(CM.can_apply(HM))
+						chromosomes.Add(CM)
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 						chromosomes += CM
```

File: code/game/machinery/computer/dna_console.dm [hunk 9]
```diff
@@ -757,9 +796,29 @@
...
+				if(LAZYLEN(chromosomes))
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
 				if(chromosomes.len)
```

File: code/game/objects/items/chromosome.dm [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+		return TRUE
+	if((synchronizer_coeff != 1) && (HM.synchronizer_coeff != -1))
+		return TRUE
+	if((power_coeff != 1) && (HM.power_coeff != -1))
+		return TRUE
+	if((energy_coeff != 1) && (HM.energy_coeff != -1))
+		return TRUE
+
+/obj/item/chromosome/proc/apply(datum/mutation/human/HM)
+	if(!HM || !HM.can_chromosome)
```
Comment (Cyberboss) [inline code/game/objects/items/chromosome.dm]:
```suggestion
 	if(!HM.can_chromosome)
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/__HELPERS/dna.dm [hunk 11]
```diff
@@ -3,6 +3,10 @@
...
-#define mutation_in_sequence(A, B) ((A) in B.mutation_index)
 #define get_gene_string(A, B) (B.mutation_index[A])
-#define get_sequence(A) (GLOB.full_sequences[A])
\ No newline at end of file
+#define get_sequence(A) (GLOB.full_sequences[A])
+
+#define get_stabilizer(A) ((A.stabilizer_coeff < 0) ? 1 : A.stabilizer_coeff)
+#define get_synchronizer(A) ((A.synchronizer_coeff < 0) ? 1 : A.synchronizer_coeff)
+#define get_power(A) ((A.power_coeff < 0) ? 1 : A.power_coeff)
+#define get_energy(A) ((A.energy_coeff < 0) ? 1 : A.energy_coeff)
```
[the remark this answers is not in the document]
Comment (optimumtact) [reply code/__HELPERS/dna.dm]: ^

**Comment by Time-Green** [comment]
Thanks for the quick review

**Comment by Jordie0608** [comment]
Conflicts

**Comment by Time-Green** [comment]
@Jordie0608 conflict fixed

**Comment by carshalash** [comment]
For some reason accessing powers that have been stored in the console dont work unless the current monkeyman in the pod has said power.

**Comment by PKPenguin321** [comment]
I somewhat dislike that these are randomized, but it's still really cool

**Comment by Time-Green** [comment]
I’ve got exams so I’ll be a little slower with fixing stuff

**Comment by Time-Green** [comment]
Fixed conflicts and the bug carsh pointed out

File: code/datums/mutations.dm [hunk 12]
```diff
@@ -31,12 +31,26 @@
 	//MUT_EXTRA - A mutation that is in the mutations tab, and can be given and taken away through though the DNA console. Has a 0 before it's name in the mutation section of the dna console
 	//MUT_OTHER Cannot be interacted with by players through normal means. I.E. wizards mutate
 
-/datum/mutation/human/New(class_ = MUT_OTHER, timer)
+
+	var/can_chromosome = 1 //can we take chromosomes? 0: CHROMOSOME_NEVER never,  1:CHROMOSOME_NONE yeah, 2: CHROMOSOME_USED no, already have one
```
Comment (Cyberboss) [inline code/datums/mutations.dm]: Should have `#define`d values

File: code/game/machinery/computer/dna_console.dm [hunk 13]
```diff
@@ -420,13 +452,23 @@
...
 				mutcolor = "bad"
+		if(HM)
+			instability *= GET_MUTATION_STABILIZER(HM)
 		temp_html += "<div class='statusDisplay'><div class='statusLine'><span class='[mutcolor]'><b>[mut_name]</b></span><small> ([alias])</small><br>"
+		temp_html += "<div class='statusDisplay'><div class='statusLine'>Instability : [round(instability)]</span><br>"
 	else
 		temp_html += "<div class='statusDisplay'><div class='statusLine'><b>[alias]</b><br>"
 	temp_html += "<div class='statusLine'>[mut_desc]<br></div>"
+	if(active && !storage_slot)
+		if(HM && HM.can_chromosome && (HM in viable_occupant.dna.mutations))
```
Comment (Cyberboss) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
		if(HM?.can_chromosome && (HM in viable_occupant.dna.mutations))
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/datums/mutations.dm (hunk 12 above)
Comment (Time-Green) [reply code/datums/mutations.dm]: they're defined in __DEFINED/DNA.dm

**Comment by Time-Green** [comment]
confllicts fixed

**Comment by Time-Green** [comment]
fixes conflicts

**Comment by carshalash** [comment]
So how much genetic damage is removed from the stabilizer chromie? I remember hearing that using it on goon removes the defect entirely.

**Comment by Time-Green** [comment]
It halves it. I’ll probably make them harder to get, just not sure yet

File: code/datums/mutations.dm (hunk 12 above)
Comment (optimumtact) [reply code/datums/mutations.dm]: so why aren't you using it

Comment (Time-Green) [reply code/datums/mutations.dm]: Oh i didnt even notice

**Comment by PKPenguin321** [comment]
Yeah, instability-halving is a bit sketchy to be honest

**Comment by Time-Green** [comment]
Yeah I know. The whole point of instability is as a limiter and letting people halve it is going to ruin the already fragile balance of genetics. 
I’ll probably make some mutations immune to it and make it lower instability by 20% instead of halved

**Comment by Time-Green** [comment]
Makes stabilizer reduce by 20% and makes it rarer.
You can only get chromosomes by activating mutations on players, so it now requires semi-social interaction
Also adds a chance of no chromosome dropping
Adds reinforcer chromosome, wich makes a mutation mutadone proof.

Generally turns chromosomes into something you dont have 200 of anymore and makes stabilizers pretty much not worth it

File: code/datums/mutations/actions.dm [hunk 14]
```diff
@@ -43,6 +52,17 @@
...
 
+/obj/effect/proc_holder/spell/aimed/firebreath/ready_projectile(obj/item/projectile/P, atom/target, mob/user, iteration)
+	if(!istype(P, /obj/item/projectile/magic/aoe/fireball))
+		return
+	var/obj/item/projectile/magic/aoe/fireball/F = P
+	switch(strength)
+		if(0 to 3)
+			F.exp_light = strength-1
+		if(4 to INFINITY)
+			F.exp_heavy = strength-4
```
Comment (ATH1909) [inline code/datums/mutations/actions.dm]: So a strength 4 firebreath mutation's exp_light radius would be _smaller_ than that of a strength 3 firebreath mutation?

**Comment by adamkad1** [comment]
when will you remove the ATGC game and make genetics activator based

**Comment by Time-Green** [comment]
@adamkad1 please elaborate

File: code/datums/mutations/actions.dm (hunk 14 above)
Comment (Time-Green) [reply code/datums/mutations/actions.dm]: Oops, good catch

**Comment by adamkad1** [comment]
> ever play on goons? back then you used to match at gc and such, but now you just make activators and inject them, then put the filled ones back in computer



> Yeah, that’s gonne be a never

makes sense

**Comment by ExcessiveUseOfCobblestone** [comment]
To-Do: Actually review this

**Comment by Time-Green** [comment]
just review the last six commits, I hardly changed anything

File: code/datums/mutations/actions.dm [hunk 15]
```diff
@@ -132,6 +141,17 @@
 			to_chat(C,"<span class='warning'>Something in front of your mouth caught fire!</span>")
 			return FALSE
 
+/obj/effect/proc_holder/spell/aimed/firebreath/ready_projectile(obj/item/projectile/P, atom/target, mob/user, iteration)
+	if(!istype(P, /obj/item/projectile/magic/aoe/fireball))
+		return
+	var/obj/item/projectile/magic/aoe/fireball/F = P
+	switch(strength)
+		if(0 to 3)
+			F.exp_light = strength-1
```
Comment (ExcessiveUseOfCobblestone) [inline code/datums/mutations/actions.dm]: nothing bad happens when the value is `-1` right?

File: code/modules/modular_computers/file_system/programs/powermonitor.dm [hunk 16]
```diff
@@ -51,7 +51,7 @@
 	if(!local_apc.terminal) //this really shouldn't happen without badminnery.
 		local_apc = null
 
-/datum/computer_file/program/power_monitor/proc/get_powernet() //keep in sync with /obj/machinery/computer/monitor's version
+/datum/computer_file/program/power_monitor/proc/GET_MUTATION_POWERnet() //keep in sync with /obj/machinery/computer/monitor's version
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/modular_computers/file_system/programs/powermonitor.dm]: please change the procname it almost looks like a define

File: code/modules/power/monitor.dm [hunk 17]
```diff
@@ -58,7 +58,7 @@
 	if(!local_apc.terminal) //this really shouldn't happen without badminnery.
 		local_apc = null
 
-/obj/machinery/computer/monitor/proc/get_powernet() //keep in sync with /datum/computer_file/program/power_monitor's version
+/obj/machinery/computer/monitor/proc/GET_MUTATION_POWERnet() //keep in sync with /datum/computer_file/program/power_monitor's version
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/power/monitor.dm]: dittos

**Comment by PKPenguin321** [comment]
You really ought to add how these work to the wiki page

**Comment by Time-Green** [comment]
Yeah you’re right, I keep forgetting to update the wiki. I’ll get on it soon enough

**Pull request merged**